src/Components/Components/src/PersistentState/PersistentServicesRegistry.cs

```diff
@@ -0,0 +1,230 @@
+    private void RestoreRegistrationsIfAvailable(PersistentComponentState state)
+    {
+        foreach (var registration in _registrations)
+        {
+            var type = ResolveType(registration.Assembly, registration.FullTypeName);
+            if (type == null)
+            {
+                continue;
+            }
+
+            var instance = _serviceProvider.GetService(type);
+            if (instance != null)
+            {
+                RestoreInstanceState(instance, type, state);
+            }
+        }
+    }
+
+    [RequiresUnreferencedCode("Calls Microsoft.AspNetCore.Components.PersistentComponentState.TryTakeFromJson(String, Type, out Object)")]
+    private static void RestoreInstanceState(object instance, Type type, PersistentComponentState state)
+    {
+        var accessors = _cachedAccessorsByType.GetOrAdd(instance.GetType(), static (runtimeType, declaredType) => new PropertiesAccessor(runtimeType, declaredType), type);
+        foreach (var (key, propertyType) in accessors.KeyTypePairs)
+        {
+            if (state.TryTakeFromJson(key, propertyType, out var result))
+            {
+                var (setter, getter) = accessors.GetAccessor(key);
+                setter.SetValue(instance, result!);
+            }
+        }
+    }
+
+    private static IPersistentComponentRegistration[] ResolveRegistrations(IEnumerable<IPersistentComponentRegistration> registrations) => [.. registrations.DistinctBy(r => (r.Assembly, r.FullTypeName)).OrderBy(r => r.Assembly).ThenBy(r => r.FullTypeName)];
+
+    private Type? ResolveType(string assembly, string fullTypeName) => _persistentServiceTypeCache.GetPersistentService(assembly, fullTypeName);
+
+    private sealed class PropertiesAccessor
+    {
+        internal const BindingFlags BindablePropertyFlags = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.IgnoreCase;
+
+        private readonly Dictionary<string, (PropertySetter, PropertyGetter)> _underlyingAccessors;
+        private readonly (string, Type)[] _cachedKeysForService;
+
+        public PropertiesAccessor([DynamicallyAccessedMembers(LinkerFlags.Component)] Type targetType, Type keyType)
+        {
+            _underlyingAccessors = new Dictionary<string, (PropertySetter, PropertyGetter)>(StringComparer.OrdinalIgnoreCase);
+
+            var keys = new List<(string, Type)>();
+            foreach (var propertyInfo in GetCandidateBindableProperties(targetType))
+            {
+                SupplyParameterFromPersistentComponentStateAttribute? parameterAttribute = null;
+                foreach (var attribute in propertyInfo.GetCustomAttributes())
+                {
+                    if (attribute is SupplyParameterFromPersistentComponentStateAttribute persistentStateAttribute)
```

Comment:
Would be great to rename this to `PersistAttribute`

Reply:
I've been thinking of calling it `[SupplyParameterFromState]`

Reply:
To me at least, any name starting `SupplyParameterFrom` doesn't sound like something you'd want to use regularly. It's not clean and concise like `[Parameter]`. 

@danroth27 What's your take? Do you aim for this to be a feature that everyone uses all the time, or is it more niche?

Reply:
`[SupplyParameterFromXxx]` is the pattern we've been using for component parameters, but this attribute isn't just for component parameters, correct? I believe we intend it to also be used from services.

Someone in the recent community standup suggest `[State]`, but I worry that doesn't really clarify what kind of state is being identified. How about `[PersistentState]`?

Reply:
I'm expecting us to discuss this in the API review process. I'm happy to bikeshed names, but I'm going to go with the ugly one for now as this will be decided on API review